### PR TITLE
feat: add macOS native capture and cursor pipeline

### PR DESCRIPTION
**Pull request by EtienneLescot**
## Description

This draft PR starts the **macOS-only** port of the native capture and cursor pipeline introduced for Windows in PR #217.

The Windows native capture PR has been merged, so this branch is now based directly on latest `upstream/main`.

The goal is to bring the same architectural separation to macOS: Electron owns recording/session orchestration, while a native ScreenCaptureKit helper owns screen/window capture, cursor exclusion, audio/webcam timing, encoding, and muxing.

## What changed

- Adds a macOS native recorder roadmap at `docs/engineering/macos-native-recorder-roadmap.md`.
- Defines the first macOS helper contract in `src/lib/nativeMacRecording.ts`.
- Adds source-id parsing helpers for ScreenCaptureKit display/window ids.
- Adds focused tests for macOS source-id parsing.
- Adds `npm run build:native:mac` as a placeholder build entrypoint.
- Documents the expected macOS helper layout in `electron/native/README.md`.
- Ignores generated macOS native helper build output and SwiftPM/Xcode state.

## Follow-up fixes from macOS testing

- Improves HUD click-through handling on macOS so dock controls, the drag handle, and popover interactions receive mouse events reliably while transparent areas still pass clicks through.
- Replaces the native Electron drag-region dependency for the HUD grab handle with an explicit pointer-driven window move IPC path.
- Fixes the language picker popover so it can be scrolled and language choices can be clicked in the transparent HUD window.
- Adds the missing editor → recorder IPC handlers used by the "Return to Recorder" confirmation flow.
- Adds editor preview support for macOS recordings with separate system-audio and microphone tracks by preparing and syncing a supplemental preview audio track for existing multi-track MP4 recordings.
- Syncs locale key structures so the macOS pre-commit/i18n check passes.

## macOS-only scope

This PR does not change Windows native capture behavior and does not add Linux native capture.

The macOS helper implementation is not shipped yet. This first step is the architecture/contract scaffold so the native helper can be added behind a clean boundary, matching the Windows WGC helper separation.

## Architecture notes

The intended macOS backend follows the same separation used by the Windows helper:

- macOS capture through ScreenCaptureKit-based native helpers.
- Windows capture through a separate WGC helper.
- Electron as the process/session coordinator.
- Renderer code consuming stable contracts rather than helper-specific process details.

## Validation run

Passed locally on Windows:

- `npm run build:native:mac` (skips successfully on non-macOS hosts)
- `npm run build-vite`
- `npx vitest run src\lib\nativeMacRecording.test.ts --pool=threads`

Previously passed before rebase:

- `npm test` (137 tests passing)

Passed locally on macOS (May 13, 2026):

- `npm run build:native:mac`
- `npm run build-vite`
- `npm run lint`
- `npm run i18n:check`
- `npm test` (139 tests passing)
- Electron smoke: source selector screens/windows + thumbnails, full-screen/window source selection, open video, open project, playback/pause/seek
- Electron macOS native recording smoke with fake mic + system audio: mic-only, system-audio-only, mic + system audio produced expected video/audio tracks

Known local validation gap:

- Camera/webcam recording was not tested on this cloud Mac.

### Capture + Launch

- [x] pre commit check on macOS
- [x] Source selector opens and lists screens/windows with thumbnails
- [x] Selecting a source works for full screen and app window
- [x] Record start/stop works from launch window
- [x] Tray “Stop Recording” works while recording
- [x] Opening existing video file works from launch window
- [x] Opening existing project works from launch window

### Audio

- [x] Mic toggle on/off works before recording
- [x] Mic device selection works
- [x] System audio toggle works across recordings
- [x] Mic-only recording works
- [x] System-audio-only recording works
- [x] Mic + system audio mix works and levels are balanced

### Editor Load + Playback

- [x] Playback, pause, seek works.
- [x] Cursor telemetry overlays correctly

### Timeline

- [x] Add/edit/remove different elements (zoom/ annotation/ trim/ speed), see how they are in playback and if they are 1:1 with exported video
- [x] Region drag/resize snaps and persists correctly
- [x] No overlap/ordering bugs on timeline items (apart from annotations)

### Project Persistence

- [x] save/ load works

### Other

- [x] Tweaking sliders/ wallpapers/ other effects.

## Summary by CodeRabbit

* **New Features**
  * Native macOS screen/window recording with pause/resume, attachable webcam, native cursor capture, and supplemental preview audio
  * HUD overlay repositioning via drag and programmatic move API
  * Cursor "clip to bounds" export/preview option and editor UI control

* **Improvements**
  * On-demand permission flows and clearer retry guidance during recording startup
  * Editable cursor overlay enabled on macOS when native cursor data present
  * Preview-audio syncing for supplemental audio tracks

* **Documentation**
  * Added macOS native recorder roadmap and native helper README

* **Chores**
  * Mac helper build/packaging script, packaging updates, and .gitignore updates

* **Localization**
  * Expanded translations and new cursor/recording strings across many locales

* **Tests**
  * Added unit tests for mac source ID parsing and cursor telemetry behaviors

[Review Change Stack](https://app.coderabbit.ai/change-stack/siddharthvaddem/openscreen/pull/573)

### CODE REVIEW

**Comment by EtienneLescot** [comment]
@siddharthvaddem I am starting to work on this ;-)

**Comment by auberginewly** [comment]
Hey @EtienneLescot! Really excited to see this PR — I've been researching exactly this problem for openscreen.

I did a deep dive into macOS cursor control for screen recording, covering:
- `SCStreamConfiguration.showsCursor = false` as the correct API
- Why Electron's `desktopCapturer` can't currently exclude the cursor
- NSCursor.currentSystem image-hashing for cursor type detection
- Coordinate mapping for Retina/multi-display setups

I'd love to contribute to the Swift helper process part. What's the best way to help right now — is there a specific piece you'd like someone to pick up?

**Comment by EtienneLescot** [comment]
> Hey @EtienneLescot! Really excited to see this PR — I've been researching exactly this problem for openscreen.
> 
> I did a deep dive into macOS cursor control for screen recording, covering:
> 
> * `SCStreamConfiguration.showsCursor = false` as the correct API
> * Why Electron's `desktopCapturer` can't currently exclude the cursor
> * NSCursor.currentSystem image-hashing for cursor type detection
> * Coordinate mapping for Retina/multi-display setups
> 
> I'd love to contribute to the Swift helper process part. What's the best way to help right now — is there a specific piece you'd like someone to pick up?

Hey @auberginewly, sure, I'll probably need help for edge cases and Retina/multi-display setups, I'll let you know when I am there.

**Comment by auberginewly** [comment]
Sounds good, ping me whenever you're ready! ☺️ @EtienneLescot

**Comment by EtienneLescot** [comment]
macOS validation note: I fixed the i18n pre-commit failure by syncing locale key structures against the English baseline. Part of that is directly related to this PR (macOS app/menu labels in common.actions.* and the launch cursor toggle labels), and part is broader catalog cleanup required because npm run i18n:check validates every locale/namespace, not only keys touched by this PR. I kept it as a separate, committable i18n cleanup so the macOS pre-commit check can legitimately stay checked.

**Comment by EtienneLescot** [comment]
@siddharthvaddem I tested it on mac, all seems ok.
I just can't test the camera.
@auberginewly if you want to test the camera that would be of help ;-)

File: electron/ipc/handlers.ts [hunk 1]
```diff
@@ -785,6 +987,134 @@ function readNativeWindowsWebcamFormat(output: string) {
...
+			proc.off("close", onClose);
+			proc.off("error", onError);
+		};
+
+		proc.stdout.on("data", onOutput);
+		proc.stderr.on("data", onOutput);
+		proc.once("close", onClose);
+		proc.once("error", onError);
+	});
+}
```
[the remark this answers is not in the document]
Comment (EtienneLescot) [reply electron/ipc/handlers.ts]: Fixed in 40ebe72. The macOS helper stdout/stderr now has a lifetime drain that appends output and emits parsed helper events; the start/stop waiters subscribe to that shared event stream instead of owning/removing the only pipe listeners.

File: electron/native/screencapturekit/Sources/OpenScreenMacOSCursorHelper/main.swift [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+}
+
+func parentElement(_ element: AXUIElement) -> AXUIElement? {
+	var value: CFTypeRef?
+	let result = AXUIElementCopyAttributeValue(element, kAXParentAttribute as CFString, &value)
+	guard result == .success else {
+		return nil
+	}
+
+	return (value as! AXUIElement)
```
[the remark this answers is not in the document]
Comment (EtienneLescot) [reply electron/native/screencapturekit/Sources/OpenScreenMacOSCursorHelper/main.swift:118]: Fixed in 40ebe72. I kept the cast safe for CoreFoundation by checking CFGetTypeID(value) against AXUIElementGetTypeID() before returning the AXUIElement, so unexpected accessibility values now fall back to nil instead of crashing.

File: electron/native/screencapturekit/Sources/OpenScreenScreenCaptureKitHelper/main.swift [hunk 3]
```diff
@@ -0,0 +1,673 @@
...
+				}
+				semaphore.wait()
+				if AVCaptureDevice.authorizationStatus(for: .audio) != .authorized {
+					throw HelperError.permissionDenied("Microphone permission is required for native microphone capture.")
+				}
+			default:
+				throw HelperError.permissionDenied("Microphone permission is required for native microphone capture.")
+			}
+		}
+	}
```
[the remark this answers is not in the document]
Comment (EtienneLescot) [reply electron/native/screencapturekit/Sources/OpenScreenScreenCaptureKitHelper/main.swift:338]: Fixed in 40ebe72. The microphone permission request now uses semaphore.wait(timeout: .now() + 30) and treats timeout or any non-authorized final status as permission denied.

File: package.json [hunk 4]
```diff
@@ -20,6 +20,7 @@
 		"format": "biome format --write .",
 		"i18n:check": "node scripts/i18n-check.mjs",
 		"preview": "vite preview",
+		"build:native:mac": "node scripts/build-macos-screencapturekit-helper.mjs",
 		"build:mac": "tsc && vite build && electron-builder --mac",
```
[the remark this answers is not in the document]
Comment (EtienneLescot) [reply package.json]: Fixed in 40ebe72. build:mac now runs npm run build:native:mac before the TypeScript/Vite/electron-builder mac packaging step.

File: scripts/build-macos-screencapturekit-helper.mjs [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+if (result.status !== 0) {
+	process.exit(result.status ?? 1);
+}
+
+fs.mkdirSync(buildDir, { recursive: true });
+fs.mkdirSync(distributableDir, { recursive: true });
+fs.copyFileSync(builtHelperPath, localHelperPath);
+fs.copyFileSync(builtHelperPath, distributablePath);
+fs.copyFileSync(builtCursorHelperPath, localCursorHelperPath);
+fs.copyFileSync(builtCursorHelperPath, distributableCursorHelperPath);
```
[the remark this answers is not in the document]
Comment (EtienneLescot) [reply scripts/build-macos-screencapturekit-helper.mjs:82]: Fixed in 40ebe72. The macOS helper build script now checks both expected Swift artifacts explicitly and prints the missing path before copy/chmod.

File: src/components/launch/LaunchWindow.tsx [hunk 6]
```diff
@@ -281,13 +282,23 @@ export function LaunchWindow() {
...
 		return () => {
 			window.electronAPI?.setHudOverlayIgnoreMouseEvents?.(false);
 		};
 	}, []);
 
+	useEffect(() => {
+		if (isLanguageMenuOpen) {
+			setHudMouseEventsEnabled(true);
+		}
+	}, [isLanguageMenuOpen, setHudMouseEventsEnabled]);
```
[the remark this answers is not in the document]
Comment (EtienneLescot) [reply src/components/launch/LaunchWindow.tsx:303]: Fixed in 40ebe72. HUD mouse passthrough is now restored symmetrically when the language menu closes and when HUD dragging ends, and repeated identical IPC calls are suppressed with a local last-value ref.

File: src/hooks/useScreenRecorder.ts [hunk 7]
```diff
@@ -677,14 +809,141 @@ export function useScreenRecorder(): UseScreenRecorderReturn {
...
+			allowAutoFinalize.current = true;
+			setRecording(true);
+			setPaused(false);
+			setElapsedSeconds(0);
+			return true;
+		} catch (error) {
+			console.error("Native macOS capture failed:", error);
+			throw error;
+		}
+	};
```
[the remark this answers is not in the document]
Comment (EtienneLescot) [reply src/hooks/useScreenRecorder.ts:954]: Fixed in 40ebe72. The native macOS start path now re-checks the countdown token after the webcam readiness wait, before starting the helper, and after helper start; stale flows stop any sidecar webcam recorder and discard the just-started native capture instead of flipping recording state.

File: src/hooks/useScreenRecorder.ts [hunk 8]
```diff
@@ -705,6 +964,27 @@ export function useScreenRecorder(): UseScreenRecorderReturn {
...
 		}
 
+		try {
+			const platform = await window.electronAPI.getPlatform();
+			if (platform === "darwin" && cursorCaptureMode === "editable-overlay") {
+				const access = await window.electronAPI.requestNativeMacCursorAccess();
+				if (!access.granted) {
+					toast.info(
+						"Allow Accessibility access for OpenScreen, then press record again to start the countdown.",
+					);
```
[the remark this answers is not in the document]
Comment (EtienneLescot) [reply src/hooks/useScreenRecorder.ts]: Fixed in 40ebe72. The Accessibility retry toast now uses editor.recording.accessibilityAllowAndRetry and that key has been added across all supported locales.

File: src/i18n/locales/tr/launch.json [hunk 9]
```diff
@@ -15,17 +15,7 @@
...
-		"noiseReduction": "Gürültü azaltma",
-		"clickToCycle": "Seviye değiştirmek için tıklayın",
-		"nrLevel": {
-			"light": "Hafif",
-			"moderate": "Orta",
-			"aggressive": "Güçlü"
-		},
-		"noiseReductionPrompt": "Daha net ses için yapay zeka destekli gürültü azaltmayı etkinleştirmek ister misiniz?",
-		"enableNoiseReductionShort": "Etkinleştir"
+		"nrLevel": {}
```
[the remark this answers is not in the document]
Comment (EtienneLescot) [reply src/i18n/locales/tr/launch.json]: Fixed in 40ebe72. Removed the unused empty nrLevel stubs from the Turkish launch/settings locale files and verified i18n parity.

File: src/hooks/useScreenRecorder.ts [hunk 10]
```diff
@@ -643,7 +775,7 @@ export function useScreenRecorder(): UseScreenRecorderReturn {
 					},
 				},
 				webcam: {
-					enabled: webcamEnabled,
+					enabled: false,
```
[the remark this answers is not in the document]
Comment (EtienneLescot) [reply src/hooks/useScreenRecorder.ts]: Fixed in 40ebe72. The native Windows request now preserves webcamEnabled so Windows webcam capture remains enabled when the HUD toggle is on.

File: src/hooks/useScreenRecorder.ts [hunk 11]
```diff
@@ -455,11 +475,115 @@ export function useScreenRecorder(): UseScreenRecorderReturn {
...
+	);
+
 	const stopRecording = useRef(() => {
 		if (nativeWindowsRecording.current) {
 			void finalizeNativeWindowsRecording(false);
 			return;
 		}
+		if (nativeMacRecording.current) {
+			void finalizeNativeMacRecording(false);
+			return;
```
[the remark this answers is not in the document]
Comment (EtienneLescot) [reply src/hooks/useScreenRecorder.ts:585]: Fixed in 40ebe72. Restart and cancel now handle native macOS sessions by finalizing/discarding through finalizeNativeMacRecording(true), mirroring the native Windows path.

File: electron/native/screencapturekit/Sources/OpenScreenMacOSCursorHelper/main.swift [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+		guard let source = CFMachPortCreateRunLoopSource(kCFAllocatorDefault, tap, 0) else {
+			return false
+		}
+
+		eventTap = tap
+		runLoopSource = source
+		CFRunLoopAddSource(CFRunLoopGetCurrent(), source, .commonModes)
+		CGEvent.tapEnable(tap: tap, enable: true)
+		return true
+	}
```
[the remark this answers is not in the document]
Comment (EtienneLescot) [reply electron/native/screencapturekit/Sources/OpenScreenMacOSCursorHelper/main.swift:51]: Fixed in 7675895. The cursor helper now handles tapDisabledByTimeout and tapDisabledByUserInput by re-enabling the stored CGEvent tap before returning, so click tracking can recover during long recording sessions.

File: electron/native/screencapturekit/Sources/OpenScreenMacOSCursorHelper/main.swift [hunk 13]
```diff
@@ -0,0 +1,258 @@
...
+
+	return nil
+}
+
+func accessibilityPointForMouse() -> CGPoint {
+	let mouse = NSEvent.mouseLocation
+	let maxY = NSScreen.screens.map { $0.frame.maxY }.max() ?? NSScreen.main?.frame.height ?? 0
+	return CGPoint(x: mouse.x, y: maxY - mouse.y)
+}
```
[the remark this answers is not in the document]
Comment (EtienneLescot) [reply electron/native/screencapturekit/Sources/OpenScreenMacOSCursorHelper/main.swift:196]: Fixed in 7675895. accessibilityPointForMouse now flips Y against the primary screen height instead of the global maxY, which keeps AX hit-testing aligned on multi-display layouts.

File: src/components/launch/LaunchWindow.tsx [hunk 14]
```diff
@@ -368,9 +405,13 @@ export function LaunchWindow() {
 			onPointerMove={(event) => {
 				const target = event.target as HTMLElement | null;
 				const shouldCapture = Boolean(target?.closest("[data-hud-interactive='true']"));
-				window.electronAPI?.setHudOverlayIgnoreMouseEvents?.(!shouldCapture);
+				setHudMouseEventsEnabled(shouldCapture);
+			}}
```
[the remark this answers is not in the document]
Comment (EtienneLescot) [reply src/components/launch/LaunchWindow.tsx:410]: Fixed in 7675895. The HUD pointer-move capture check now keeps mouse events enabled whenever the language menu is open, so crossing transparent space between the trigger and the portaled menu no longer makes the picker unclickable.

File: src/i18n/locales/tr/settings.json [hunk 15]
```diff
@@ -185,15 +194,5 @@
...
-		"noiseReduction": "Gürültü Azaltma",
-		"level": "Seviye",
-		"nrLevel": {
-			"light": "Hafif",
-			"moderate": "Orta",
-			"aggressive": "Güçlü"
-		},
-		"nrDescription": "Yapay zeka destekli gürültü azaltma arka plan gürültüsünü temizler. Daha yüksek seviyeler daha agresiftir ancak ses kalitesini etkileyebilir."
-	}
+	"audio": {}
```
[the remark this answers is not in the document]
Comment (EtienneLescot) [reply src/i18n/locales/tr/settings.json]: Fixed in 7675895. Removed the orphaned empty audio object from tr/settings.json and re-ran the i18n check.

**Comment by auberginewly** [comment]
> @siddharthvaddem I tested it on mac, all seems ok. I just can't test the camera. @auberginewly if you want to test the camera that would be of help ;-)

ok

**Comment by EtienneLescot** [comment]
@auberginewly except for camera testing, all is good on my side, were you able to do a camera test?
@siddharthvaddem do you know contributors that have mac + webcam?
Aside from that cam checks others checks are ok.

**Comment by siddharthvaddem** [comment]
@EtienneLescot Do you mean an external webcam or the one built in with mac. I do not have an external webcam unfortunately.

Tagging @FabLrc @marcgabe15 for visibility

**Comment by auberginewly** [comment]
Tested on macOS — here's what I found:

**✅ Camera:** Works fine, no issues.

**🐛 Cursor goes off-screen (external display):** When recording on an external monitor, the cursor overlay moves outside the screen boundary. Not sure if this also happens on a single display — could be a coordinate mapping issue specific to multi-display setups. (screenshot below)


image

**Comment by EtienneLescot** [comment]
@auberginewly Thank you for the cam test !! Much appreciated!
Concerning the cursor overlay moving outside the screen boundary.
I am no sure to understand exactly the issue. I have only one screen so I can't reproduce. But your screenshot does not seem wrong...
Could you explain a bit more, maybe it is me that do not understand the point.

**Comment by auberginewly** [comment]
> @auberginewly Thank you for the cam test !! Much appreciated! Concerning the cursor overlay moving outside the screen boundary. I am no sure to understand exactly the issue. I have only one screen so I can't reproduce. But your screenshot does not seem wrong... Could you explain a bit more, maybe it is me that do not understand the point.

In the post-recording editor, the magnified cursor overlay overflows beyond the video canvas boundary into the Openscreen background area. The cursor overlay should be clipped to the video canvas — when the cursor reaches the edge of the recorded footage, the overlay should stop at the boundary rather than spilling into the background region, to ensure a clean viewing experience.

**Comment by auberginewly** [comment]
> > @auberginewly Thank you for the cam test !! Much appreciated! Concerning the cursor overlay moving outside the screen boundary. I am no sure to understand exactly the issue. I have only one screen so I can't reproduce. But your screenshot does not seem wrong... Could you explain a bit more, maybe it is me that do not understand the point.
> 
> In the post-recording editor, the magnified cursor overlay overflows beyond the video canvas boundary into the Openscreen background area. The cursor overlay should be clipped to the video canvas — when the cursor reaches the edge of the recorded footage, the overlay should stop at the boundary rather than spilling into the background region, to ensure a clean viewing experience.

Found the root cause — in projectNativeCursorToStage, the stage coordinate isn't clamped to the video canvas bounds after the global projection. Would it be ok if I submit a fix for this?

**Comment by EtienneLescot** [comment]
@auberginewly of course, you can make a PR on my branch !
Simply i am not sure to understant the use case. on your screenshot I see a cursor visible in the padding area.
Not sure that it is a "bad behavior" you know what I mean ? I might be wrong but, to me it seems cleaner this way instead of having a cropped cursor when user show elements at the edges.

**Comment by auberginewly** [comment]
> @auberginewly of course, you can make a PR on my branch !
> Simply i am not sure to understant the use case. on your screenshot I see a cursor visible in the padding area.
> Not sure that it is a "bad behavior" you know what I mean ? I might be wrong but, to me it seems cleaner this way instead of having a cropped cursor when user show elements at the edges.

What I mean is: when the cursor is enlarged, part of it overflows outside the video canvas. My suggestion is to apply a clip mask to the video canvas, so the portion of the cursor that goes outside is hidden — not removing the cursor entirely. This way the cursor still appears naturally at the edges, just clipped by the canvas boundary, like CSS overflow: hidden.

I'll start working on a fix for this now.

**Comment by EtienneLescot** [comment]
> @EtienneLescot Do you mean an external webcam or the one built in with mac. I do not have an external webcam unfortunately.
> 
> Tagging @FabLrc @marcgabe15 for visibility

@siddharthvaddem I meant whatever camere built-in or external.
it seems like @auberginewly tested and that it was ok though.

Concerning the cursor overflowing in the padding area, I find that it is better like that (with overflow). 
Also this is the way it is working in Windows and it this ok this way.
What do you guys think?

Tagging @FabLrc @marcgabe15 for visibility

**Comment by auberginewly** [comment]
> > @EtienneLescot Do you mean an external webcam or the one built in with mac. I do not have an external webcam unfortunately.
> > Tagging @FabLrc @marcgabe15 for visibility
> 
> @siddharthvaddem I meant whatever camere built-in or external. it seems like @auberginewly tested and that it was ok though.
> 
> Concerning the cursor overflowing in the padding area, I find that it is better like that (with overflow). Also this is the way it is working in Windows and it this ok this way. What do you guys think?
> 
> Tagging @FabLrc @marcgabe15 for visibility

cursor overflow in the padding area — I understand you prefer keeping the overflow behavior, and it aligns with Windows' default behavior. No issue there.

Quick update on my side: I cleaned up unnecessary mouse movement trails across multi-monitor setups and fixed the cursor jump/disconnect issue when moving between displays.

One enhancement suggestion: could we make "cursor constrained to canvas / allowed to overflow" a configurable option?

This gives users flexibility:
Overflow allowed — freeform, matches Windows behavior (your preference)
Clamped to canvas — cleaner, better for presentations/screencasts
Implementation-wise, this would just need a conditional check against canvas bounds (e.g., overflow: visible vs clip behavior). If the direction looks good, I can branch off yours and open a PR to implement this.

What do you think?

**Comment by EtienneLescot** [comment]
> > > @EtienneLescot Do you mean an external webcam or the one built in with mac. I do not have an external webcam unfortunately.
> > > Tagging @FabLrc @marcgabe15 for visibility
> > 
> > 
> > @siddharthvaddem I meant whatever camere built-in or external. it seems like @auberginewly tested and that it was ok though.
> > Concerning the cursor overflowing in the padding area, I find that it is better like that (with overflow). Also this is the way it is working in Windows and it this ok this way. What do you guys think?
> > Tagging @FabLrc @marcgabe15 for visibility
> 
> cursor overflow in the padding area — I understand you prefer keeping the overflow behavior, and it aligns with Windows' default behavior. No issue there.
> 
> Quick update on my side: I cleaned up unnecessary mouse movement trails across multi-monitor setups and fixed the cursor jump/disconnect issue when moving between displays.
> 
> One enhancement suggestion: could we make "cursor constrained to canvas / allowed to overflow" a configurable option?
> 
> This gives users flexibility: Overflow allowed — freeform, matches Windows behavior (your preference) Clamped to canvas — cleaner, better for presentations/screencasts Implementation-wise, this would just need a conditional check against canvas bounds (e.g., overflow: visible vs clip behavior). If the direction looks good, I can branch off yours and open a PR to implement this.
> 
> What do you think?

As for me, I am ok with the option solution but we should also have that checkbox option in windows then.
I let you make the PR so that I can review it.

**Comment by auberginewly** [comment]
there's a little bug and i'm trying to fix it

**Comment by siddharthvaddem** [comment]
everything looks good on my end - one nit is that the cursor type does not change between a pointer to insert text/ resize/ drag and so on, it remains the default so I am wondering if that is the intended behavior? If so - should we think about making that dynamic? and also support for custom cursors https://sweezy-cursors.com/

let me check if this is a regression from my local setup (not sure what I am doing wrong, but it seems like the cursor remains the default pointer type)

**Comment by auberginewly** [comment]
现在我是已经把光标 overflow 给删掉了 修了多余运动轨迹采集的 bug 我来加一个可选 overflow or not 的开关

After more testing I've reconsidered the cursor-overflow approach:

Removed the overflow clipping. The hard clip-path/canvas clipping that confined the cursor to the video bounds is gone — clipping the native cursor at the edge looked worse than letting it sit slightly outside in some cases.
Fixed the real root cause: redundant motion-trail sampling. The phantom trail wasn't really an overflow problem — it came from stale out-of-bounds samples being collected and fed into smoothing. That's now fixed at the sampling layer, so the trail/ghosting is gone without needing to clip.
Next: make overflow opt-in. I'll add an optional "clip cursor to canvas / allow overflow" toggle so the behavior is configurable instead of hard-coded, rather than forcing one choice on everyone.
Will push the toggle commit shortly.

**Comment by auberginewly** [comment]
@siddharthvaddem @EtienneLescot

I've added a set of macOS cursor fixes + an optional toggle on top of this native capture/cursor pipeline. The PR is opened against Etienne's fork: [EtienneLescot/openscreen#2](https://github.com/EtienneLescot/openscreen/issues/2)

Scope: multi-display ghost cursor/trail, cursor-overflow clipping, preview/export size normalization, lost rounded corners & zoom over-crop fixes, an optional "Clip to Canvas" toggle, and i18n for the cursor settings section across 11 locales.

Plan: merge into @EtienneLescot's feat/macos-native-capture-pipeline branch first, then it lands in main together with this #573. Done and ready for review — thanks! 🙏

**Comment by siddharthvaddem** [comment]
feel free to tag me again when everything has been cleared out and is ready to merge

>everything looks good on my end - one nit is that the cursor type does not change between a pointer to insert text/ resize/ drag and so on, it remains the default so I am wondering if that is the intended behavior? If so - should we think about making that dynamic? and also support for custom cursors https://sweezy-cursors.com/

let me check if this is a regression from my local setup (not sure what I am doing wrong, but it seems like the cursor remains the default pointer type)

and putting this here again just in case it got missed in the thread

**Comment by EtienneLescot** [comment]
> feel free to tag me again when everything has been cleared out and is ready to merge
> 
> > everything looks good on my end - one nit is that the cursor type does not change between a pointer to insert text/ resize/ drag and so on, it remains the default so I am wondering if that is the intended behavior? If so - should we think about making that dynamic? and also support for custom cursors https://sweezy-cursors.com/
> 
> let me check if this is a regression from my local setup (not sure what I am doing wrong, but it seems like the cursor remains the default pointer type)
> 
> and putting this here again just in case it got missed in the thread

@siddharthvaddem the cursor is changing correctly on my end, strange, what is your macOS version?

**Comment by auberginewly** [comment]
> > feel free to tag me again when everything has been cleared out and is ready to merge
> > > everything looks good on my end - one nit is that the cursor type does not change between a pointer to insert text/ resize/ drag and so on, it remains the default so I am wondering if that is the intended behavior? If so - should we think about making that dynamic? and also support for custom cursors https://sweezy-cursors.com/
> > 
> > 
> > let me check if this is a regression from my local setup (not sure what I am doing wrong, but it seems like the cursor remains the default pointer type)
> > and putting this here again just in case it got missed in the thread
> 
> @siddharthvaddem the cursor is changing correctly on my end, strange, what is your macOS version?

is alright on my end, my macOS version is 26.4
image

**Comment by EtienneLescot** [comment]
@siddharthvaddem I just merged @auberginewly contribution. 
Seems all ok. 
Should be good to merge!

**Review by siddharthvaddem — APPROVED** [review]
